driver/sigmoid_focal_loss_driver.hpp

```diff
@@ -0,0 +1,797 @@
+        Tcheck powPt   = pow(1 - pT, gamma);
+        Tcheck alpha_t = alpha * t + (1 - alpha) * (1 - t);
+
+        if(dinput)
+        {
+            Tcheck dpdi      = exp(-i) / pow(1 + exp(-i), 2);
+            Tcheck dcelossdi = (-t / p + (1 - t) / (1 - p)) * dpdi;
+            Tcheck dpowptdi  = gamma * pow(1 - pT, gamma - 1) * (1 - 2 * t) * dpdi;
+
+            // L = ce_loss * pow_pt => dL/di = dceloss/di * pow_pt + ce_loss * dpowpt/di
+            Tcheck dLdi = dcelossdi * powPt + ceLoss * dpowptdi;
+            Tcheck grad = dO * dLdi;
+
+            if(alpha >= 0)
+            {
+                grad *= alpha_t;
+            }
+            dinput[dinput_tv.get_tensor_view_idx(idx)] = static_cast<Tcheck>(grad);
+        }
+
+        if(dtarget)
+        {
+            Tcheck dcelossdt = -log(p) + log(1 - p);
+            Tcheck dpowptdt  = gamma * pow(1 - pT, gamma - 1) * (1 - 2 * p);
+            // L = ce_loss * pow_pt => dL/dt = dceloss/dt * pow_pt + ce_loss * dpowpt/dt
+            Tcheck dLdt       = dcelossdt * powPt + ceLoss * dpowptdt;
+            Tcheck gradTarget = dO * dLdt;
+
+            if(alpha >= 0)
+            {
+                // alpha_t * dL/dt + dalpha_t/dt * dL
+                gradTarget = alpha_t * dLdt + (2 * alpha - 1) * ceLoss * powPt;
+            }
+            dtarget[dtarget_tv.get_tensor_view_idx(idx)] = static_cast<Tcheck>(gradTarget);
+        }
+    }
+}
+
+template <typename Tgpu, typename Tcheck>
+void mloSigmoidFocalLossFwdRunHost(Tgpu* input,
+                                   miopenTensorDescriptor_t inputDesc,
+                                   Tgpu* target,
+                                   miopenTensorDescriptor_t targetDesc,
+                                   Tcheck* workspaceHost,
+                                   Tcheck* outputHost,
+                                   float alpha   = 0.25,
+                                   float gamma   = 2,
+                                   float divisor = 1)
+{
+    auto input_tv    = miopen::get_inner_expanded_tv<5>(miopen::deref(inputDesc));
+    auto target_tv   = miopen::get_inner_expanded_tv<5>(miopen::deref(targetDesc));
+    size_t inputSize = miopen::deref(inputDesc).GetElementSize();
+
+    for(size_t id = 0; id < inputSize; ++id)
+    {
+        tensor_layout_t<5> idx(input_tv, id);
+
+        Tcheck i = static_cast<Tcheck>(input[input_tv.get_tensor_view_idx(idx)]);
+        Tcheck t = static_cast<Tcheck>(target[target_tv.get_tensor_view_idx(idx)]);
+
+        Tcheck sig    = 1 / (1 + exp(-i));
+        Tcheck ceLoss = -(t * log(sig) + (1 - t) * log(1 - sig));
+        Tcheck sigT   = sig * t + (1 - sig) * (1 - t);
+        Tcheck loss   = ceLoss * pow(1 - sigT, gamma);
+
+        if(alpha >= 0)
+        {
+            Tcheck alphaT = alpha * t + (1 - alpha) * (1 - t);
+            loss          = alphaT * loss;
+        }
+
+        workspaceHost[id] = static_cast<Tcheck>(loss / divisor);
```

Comment:
You don't need an external workspace for host functions - you are free to allocate memory right in this function (which is usually not possible or even does not make much sense on GPU).

But actually you don't need a workspace for CPU reduction, it's not a parallel algorithm without external synchronization.
```suggestion
        outputHost[0] += static_cast<Tcheck>(loss / divisor);
```
Just don't forget to initialize `outputHost[0]` with `0.0f`

Reply:
I declare workspaceHost in the driver as this allow us to compare the workspace in CPU with GPU before executing reduction. 
Additionally,  `outputHost[0] += static_cast<Tcheck>(loss / divisor)` results in verification fail as we use reduction algo in GPU. To avoid verification fail, I have to mimic the same reduction algo in CPU check function.

Reply:
@BuiChiTrung 
> I declare workspaceHost in the driver as this allow us to compare the workspace in CPU with GPU before executing reduction.

Basically workspace content is not specified and can't be compared. We can use multiple implementations for the same algorithm, and we can't do this check in common case and since this CPU implementation is used for that common case, it can't contain algorithm specific features.

> Additionally, `outputHost[0] += static_cast<Tcheck>(loss / divisor)` results in verification fail as we use reduction algo in GPU. To avoid verification fail, I have to mimic the same reduction algo in CPU check function.

Mathematically it must be the same, but since there are FP rounding errors and other FP stuff, it may result into bigger error, but still, it must work. If the error is too big, then the algorithm is not applicable and can be used.

Reply:
I have updated the verify method in CPU to naive accumulation